File: 169-majority-element/majority-element.java
//apna colege article 
class Solution {
    public int majorityElement(int[] nums) {
        int n=nums.length;
        int count=1;
        int majority=nums[0];//assuming the first element as candidate


        //phase 1:find potential majority candidate
        for(int i=0;i<n;i++){
            if(nums[i]==majority){
                count++;
            }
            else{
                count--;
                if(count==0){
                    majority=nums[i];
                    count=1;
                }
            }
        }

        //phase 2 verify the candidate
        count=0;
        for(int i=0;i<n;i++){
            if(nums[i]==majority){
                count++;
            }
        }

        //return the element if it occur more then  n/2
        if(count>=n/2){
            return majority;
        }else{
            return -1;
        }



    }
}

File: 169-majority-element/majority-element.py
# apna college article
class Solution:
    def majorityElement(self, nums):
        n = len(nums)
        count = 1
        majority = nums[0]

        # phase 1 find candidate
        for i in range(1, n):
            if nums[i] == majority:
                count = count + 1
            else:
                count = count - 1
                if count == 0:
                    majority = nums[i]
                    count = 1

        # phase 2 verify candidate
        if nums.count(majority) > n // 2:
            return majority
        else:
            return -1


if __name__ == "__main__":
    nums = [2,2,1,1,2,2,2]

    Sol = Solution()
    result = Sol.majorityElement(nums)

    if result != -1:
        print(result)
    else:
        print(-1)

        
        